src/server/services/pipeline-in-progress/processData.js

```diff
@@ -0,0 +1,169 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+const moment = require('moment');
+const { log } = require('../../log');
+const Reading = require('../../models/Reading');
+const handleCumulativeReset = require('./handleCumulativeReset');
+
+
+/**
+ * Handle all data, assume that the first row is the first reading (skip this row).
+ * @Example
+ * 	row 0: reading #0
+ *  row 1: reading #1
+ * 	row 2: reading #2
+ * => reading #1 = row 1
+ *    reading #2 = row 2
+ *    reading #0 may be the cumulative value from unknown readings that may or may not have been inserted before
+ * @param {object[[]]} rows
+ * @param {number} meterID meter id being input
+ * @param {boolean} isCumulative true if the data is cumulative
+ * @param {boolean} cumulativeReset true if the cumulative data can reset
+ * @param {string} resetStart a string representation of the start time a cumulativeReset may occur after. The default resetStart time is 0:00:00
+ * @param {string} resetEnd a string representation of the end time a cumulativeReset may occur before. The default resetEnd time is 23:59:99
+ * @param readingRepetition value is 1 if reading is not duplicated. 2 if repeated twice and so on (E-mon D-mon meters)
+ * @param {boolean} onlyEndTime true if the data only has an endTimestamp
+ * @param {number} Tgap the allowed time variation in milliseconds that a gap may occur between two readings 
+ * @param {number} Tlen the allowed time variation in milliseconds that two readings may deviate from each other
+ * @param {dict} conditionSet used to validate readings (minVal, maxVal, minDate, maxDate, interval, maxError)
+ */
+
+function processData(rows, meterID, isCumulative, cumulativeReset, resetStart = '0:00:00.000', 
+					resetEnd = '23:59:99.999', readingRepetition, onlyEndTime = false, Tgap, Tlen, conditionSet) {
+
+	// If processData is succesfully finished then return result = [R0, R1, R2...RN]
+	const result = [];
+
+	let errMsg = '';
+
+	let meterReading = 0;
+	let meterReading1 = 0;
+	let meterReading2 = 0;
+
```

Comment:
The recently merged PR #632 added attributes to the meters table. They include reading, start_timestamp & end_timestamp. These three values are supposed to represent the last reading stored for that meter from the pipeline. As such, they should be the first value used as the previous reading for the next value(s) that come to the pipeline for that meter. The last value processed by the pipeline should always be stored back to the DB so it is ready for the next batch. Note the reading for cumulative data is the raw, cumulative value so it still works for the next time. (The value in the readings table will be the net usages for that time frame so it will differ.) The default value when a meter is created sets the value to 0, 0001-01-01 00:00:00, 0001-01-01 00:00:00. The plan was that such a date/time (so far back in time) would mean the pipeline would still work correctly for the first point. It will generate a warning due to the time gap and not use the first ever cumulative value due to that gap but that is what is desired. Note the Meter.js model should now return these values so it can be used to get these three values from the DB and store new ones back. We really should put them into Redux with other meter state but we can skip that for now if you want. Just leave a TODO if you do that. Let  me know if this is not clear or the default values do no work as hoped.

Reply:
I think I've completed this. I've left a TODO for the redux state since I'm not sure how to go about doing this.